### PR TITLE
Refactor backoff logic by creating a new type of LimitedBackOff

### PR DESCRIPTION
**Pull request by abrahamko**
No behavior changes.
Wrap all backoff logic in a separate type implementing the BackOff interface.

#### Test coverage
- [x] This PR includes new unit tests to go with the code changes
#### Documentation
- [x] This PR does not require updating any documentation

### CODE REVIEW

File: cmd/secrets-provider/main.go [hunk 1]
```diff
@@ -51,27 +52,21 @@ func main() {
 		printErrorAndExit(messages.CSPFK009E)
 	}
 
-	constantBackOff := backoff.NewConstantBackOff(time.Duration(secretsConfig.RetryIntervalSec) * time.Second)
-	count := 0
+	limitedBackOff := utils.NewLimitedBackOff(
+		time.Duration(secretsConfig.RetryIntervalSec)*time.Second,
```
[the remark this answers is not in the document]
Comment (abrahamko) [reply cmd/secrets-provider/main.go:56]: `go fmt` keep removing these spaces.
Are you sure we want to do it despite that?

[the remark this answers is not in the document]
Comment (orenbm) [reply cmd/secrets-provider/main.go:56]: that's odd. i guess we can live with it. please see if you can configure `go fmt` to format with spaces. if not, leave it.

File: pkg/utils/limited_backoff.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"github.com/cenkalti/backoff"
+	"time"
+)
+
+type limitedBackOff struct {
+	Interval   time.Duration
+	RetryLimit int
+	retryCount int
```
[the remark this answers is not in the document]
Comment (orenbm) [reply pkg/utils/limited_backoff.go:15]: ok thanks

File: pkg/utils/limited_backoff_test.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
+package utils
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/cenkalti/backoff"
+
+	. "github.com/smartystreets/goconvey/convey"
```
Comment (orenbm) [inline pkg/utils/limited_backoff_test.go:10]: these 2 can go together as they are both external

File: cmd/secrets-provider/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (abrahamko) [reply cmd/secrets-provider/main.go:56]: This is intentional.
See: https://github.com/golang/go/issues/1206#issuecomment-66052999

[the remark this answers is not in the document]
Comment (orenbm) [reply cmd/secrets-provider/main.go:56]: thanks, go for it

**Review by orenbm — APPROVED** [review]

**Pull request merged**